File: leetcode/15_3_sum/Solution_210901.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        final boolean USED = false;
        Set<List<Integer>> answers = new HashSet<>();
        for (int i = 0; i < nums.length - 2; i++) {
            int pivot = nums[i];
            int target = 0 - pivot;
            Map<Integer, Boolean> hashmap = new HashMap<>();
            for (int j = i + 1; j < nums.length; j++) {
                int curNum = nums[j];
                if (j == i + 1) {
                    hashmap.put(curNum, !USED);
                } else {
                    int complement = target - curNum;
                    if (hashmap.containsKey(complement) && hashmap.get(complement) != USED) {
                        hashmap.put(complement, USED);
                        hashmap.put(curNum, USED);
                        List<Integer> triplet = Arrays.asList(pivot, complement, curNum);
                        Collections.sort(triplet);
                        answers.add(triplet);
                    } else if (!hashmap.containsKey(curNum)) {
                        hashmap.put(curNum, !USED);
                    }
                }
            }
        }
        for (List<Integer> l : answers) {
            // System.out.println(l);
        }
        return new ArrayList<>(answers);
    }

    public static void main(String... args){
        Solution solution = new Solution();
        int[] input = new int[]{-1,0,1,2,-1,-4};
        List<List<Integer>> answer = solution.threeSum(input);
        System.out.println(answer);
        input = new int[]{0,0,0};
        answer = solution.threeSum(input);
        System.out.println(answer);
        input = new int[]{1,1,-2};
        answer = solution.threeSum(input);
        System.out.println(answer);
    }
}

File: leetcode/15_3_sum/solution_210901.py
# Basic imports --------------------------------------------
from __future__ import annotations                         
import sys                                                 
# 파이썬 기본 재귀 limit이 1000이라고 함 --> 10^6으로 manual하게 설정
sys.setrecursionlimit(10**6)
from os.path import dirname, abspath, basename, normpath   
root = abspath(__file__)                                   
while basename(normpath(root)) != 'algo_practice':           
    root = dirname(root)                                   
sys.path.append(root)                                      
from utils.Tester import Tester, Logger                    
logger = Logger(verbose=False)                             
import pprint
pp = pprint.PrettyPrinter()
# ----------------------------------------------------------

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        USED = True
        answers = set()
        for i in range(len(nums)-2):
            pivot = nums[i]
            target = 0 - pivot
            hashmap = dict()
            for j in range(i+1, len(nums)):
                if j == i+1:
                    hashmap[nums[j]] = not USED
                else:
                    complement = target - nums[j]
                    if complement in hashmap and hashmap[complement] is not USED:
                        hashmap[complement] = USED
                        hashmap[nums[j]] = USED
                        answers.add(tuple(sorted([pivot, complement, nums[j]])))
                    elif nums[j] not in hashmap:
                        hashmap[nums[j]] = not USED
        return [list(a) for a in answers]
        
            
if __name__ == '__main__':
    sol = Solution()

    test_cases = [
        ([[-1,0,1,2,-1,-4]], [[-1,-1,2],[-1,0,1]]),
        ([[]], []),
        ([[0]], []),
    ]

    Tester.factory(test_cases, func=lambda input: sol.threeSum(*input)).run(unordered_output=True)
